inst/cpp_generation/state_types.hpp.in

```diff
@@ -0,0 +1,104 @@
+#pragma once
+
+#include "types.hpp"
+#include "parameters.hpp"
+#include "state_space.hpp"
+#include "model_variants.hpp"
+
+namespace leapfrog {
+
+namespace {
+using Eigen::Sizes;
+using Eigen::TensorFixedSize;
+}
+
+template<typename ModelVariant, typename real_type>
+struct BaseModelState {
+  TensorFixedSize <real_type, Sizes<pAG<ModelVariant>, NS<ModelVariant>>> p_total_pop;
+  TensorFixedSize <real_type, Sizes<pAG<ModelVariant>, NS<ModelVariant>>> p_total_pop_natural_deaths;
+  TensorFixedSize <real_type, Sizes<pAG<ModelVariant>, NS<ModelVariant>>> p_hiv_pop;
+  TensorFixedSize <real_type, Sizes<pAG<ModelVariant>, NS<ModelVariant>>> p_hiv_pop_natural_deaths;
+  TensorFixedSize <real_type, Sizes<hDS<ModelVariant>, hAG<ModelVariant>, NS<ModelVariant>>> h_hiv_adult;
+  TensorFixedSize <real_type, Sizes<hTS<ModelVariant>, hDS<ModelVariant>, hAG<ModelVariant>, NS<ModelVariant>>>
+      h_art_adult;
+  real_type births;
+  TensorFixedSize <real_type, Sizes<hDS<ModelVariant>, hAG<ModelVariant>, NS<ModelVariant>>> h_hiv_deaths_no_art;
+  TensorFixedSize <real_type, Sizes<pAG<ModelVariant>, NS<ModelVariant>>> p_infections;
+  TensorFixedSize <real_type, Sizes<hTS<ModelVariant>, hDS<ModelVariant>, hAG<ModelVariant>, NS<ModelVariant>>>
+      h_hiv_deaths_art;
+  TensorFixedSize <real_type, Sizes<hDS<ModelVariant>, hAG<ModelVariant>, NS<ModelVariant>>> h_art_initiation;
+  TensorFixedSize <real_type, Sizes<pAG<ModelVariant>, NS<ModelVariant>>> p_hiv_deaths;
+
+  BaseModelState(const Parameters<ModelVariant, real_type> &pars) {
```

Comment:
I moved this here as part of this PR when I was thinking I would do the output and input generation together. But I will add the generation part of this in another PR